File: rce_hawkeye/crawler.py
"""
网页爬虫模块
"""

import re
import asyncio
from typing import List, Dict, Set, Optional, Tuple
from dataclasses import dataclass, field
from urllib.parse import urljoin, urlparse, parse_qs, urlencode, unquote
import aiohttp
from bs4 import BeautifulSoup


@dataclass
class CrawledPage:
    url: str
    status_code: int
    content: str
    links: List[str] = field(default_factory=list)
    forms: List[Dict] = field(default_factory=list)
    parameters: Dict[str, str] = field(default_factory=dict)


@dataclass
class FormInfo:
    action: str
    method: str
    inputs: Dict[str, str]


class WebCrawler:
    """网页爬虫"""
    
    def __init__(
        self,
        max_depth: int = 2,
        max_pages: int = 100,
        timeout: int = 10,
        concurrent: int = 5,
        user_agent: str = "RCE-HawkEye/0.0.2"
    ):
        self.max_depth = max_depth
        self.max_pages = max_pages
        self.timeout = timeout
        self.concurrent = concurrent
        self.user_agent = user_agent
        
        self._visited: Set[str] = set()
        self._results: List[CrawledPage] = []
        self._session: Optional[aiohttp.ClientSession] = None
        self._semaphore: Optional[asyncio.Semaphore] = None
        self._all_params: Dict[str, Set[str]] = {}
    
    def _normalize_url(self, url: str) -> str:
        """标准化URL"""
        parsed = urlparse(url)
        return f"{parsed.scheme}://{parsed.netloc}{parsed.path}".rstrip('/')
    
    def _get_base_url(self, url: str) -> str:
        """获取基础URL"""
        parsed = urlparse(url)
        return f"{parsed.scheme}://{parsed.netloc}"
    
    def _extract_links(self, base_url: str, html: str) -> List[str]:
        """从HTML中提取链接"""
        links = set()
        
        try:
            soup = BeautifulSoup(html, 'html.parser')
            
            for tag in soup.find_all(['a', 'link', 'script', 'img', 'iframe', 'frame', 'form', 'area', 'base', 'source']):
                attr = None
                if tag.name in ['a', 'link', 'area']:
                    attr = 'href'
                elif tag.name in ['script', 'img', 'iframe', 'frame', 'source']:
                    attr = 'src'
                elif tag.name == 'form':
                    attr = 'action'
                elif tag.name == 'base':
                    attr = 'href'
                
                if attr and tag.get(attr):
                    href = tag.get(attr, '')
                    if href and not href.startswith(('#', 'javascript:', 'mailto:', 'tel:', 'data:')):
                        full_url = urljoin(base_url, href)
                        links.add(full_url)
            
            patterns = [
                r'(?:href|src|action|data-url|data-href)\s*=\s*["\']([^"\']+)["\']',
                r'window\.location\s*=\s*["\']([^"\']+)["\']',
                r'window\.open\s*\(\s*["\']([^"\']+)["\']',
                r'location\.href\s*=\s*["\']([^"\']+)["\']',
            ]
            
            for pattern in patterns:
                for match in re.finditer(pattern, html, re.IGNORECASE):
                    href = match.group(1)
                    if href and not href.startswith(('#', 'javascript:', 'mailto:', 'tel:', 'data:')):
                        full_url = urljoin(base_url, href)
                        links.add(full_url)
            
            url_pattern = r'(?:https?://[^\s<>"\'\)]+)'
            for match in re.finditer(url_pattern, html):
                links.add(match.group(0))
        
        except Exception:
            pass
        
        return list(links)
    
    def _extract_forms(self, base_url: str, html: str) -> List[Dict]:
        """从HTML中提取表单"""
        forms = []
        
        try:
            soup = BeautifulSoup(html, 'html.parser')
            
            for form in soup.find_all('form'):
                action = form.get('action', '')
                if action:
                    action = urljoin(base_url, action)
                else:
                    action = base_url
                
                method = form.get('method', 'GET').upper()
                
                inputs = {}
                for input_tag in form.find_all(['input', 'textarea', 'select', 'button']):
                    name = input_tag.get('name', '')
                    if name:
                        value = input_tag.get('value', '')
                        inputs[name] = value
                
                for button in form.find_all('button'):
                    name = button.get('name', '')
                    if name:
                        value = button.get('value', '')
                        inputs[name] = value
                
                forms.append({
                    'action': action,
                    'method': method,
                    'inputs': inputs
                })
            
            ajax_patterns = [
                r'\$\.ajax\s*\(\s*\{[^}]*url\s*:\s*["\']([^"\']+)["\']',
                r'\$\.post\s*\(\s*["\']([^"\']+)["\']',
                r'\$\.get\s*\(\s*["\']([^"\']+)["\']',
                r'fetch\s*\(\s*["\']([^"\']+)["\']',
                r'axios\.(?:get|post|put|delete)\s*\(\s*["\']([^"\']+)["\']',
            ]
            
            for pattern in ajax_patterns:
                for match in re.finditer(pattern, html, re.IGNORECASE):
                    url = match.group(1)
                    if url:
                        full_url = urljoin(base_url, url)
                        parsed = urlparse(full_url)
                        ajax_params = {}
                        if parsed.query:
                            query_params = parse_qs(parsed.query)
                            for key, values in query_params.items():
                                ajax_params[key] = values[0] if values else ""
                        
                        if ajax_params:
                            forms.append({
                                'action': full_url.split('?')[0],
                                'method': 'GET' if 'get' in pattern.lower() else 'POST',
                                'inputs': ajax_params
                            })
        
        except Exception:
            pass
        
        return forms
    
    def _extract_parameters_from_url(self, url: str) -> Dict[str, str]:
        """从URL中提取参数"""
        params = {}
        parsed = urlparse(url)
        
        if parsed.query:
            query_params = parse_qs(parsed.query)
            for key, values in query_params.items():
                params[key] = values[0] if values else ""
        
        return params
    
    def _extract_parameters_from_html(self, html: str, base_url: str) -> Dict[str, str]:
        """从HTML内容中提取所有可能的参数名"""
        params = {}
        
        try:
            soup = BeautifulSoup(html, 'html.parser')
            
            for input_tag in soup.find_all(['input', 'textarea', 'select']):
                name = input_tag.get('name', '')
                if name:
                    value = input_tag.get('value', '')
                    params[name] = value
            
            for tag in soup.find_all(attrs={'name': True}):
                name = tag.get('name', '')
                if name and name not in params:
                    params[name] = tag.get('value', '')
            
            patterns = [
                r'name\s*=\s*["\']([a-zA-Z_][a-zA-Z0-9_]*)["\']',
                r'data-param\s*=\s*["\']([a-zA-Z_][a-zA-Z0-9_]*)["\']',
                r'data-name\s*=\s*["\']([a-zA-Z_][a-zA-Z0-9_]*)["\']',
                r'var\s+([a-zA-Z_][a-zA-Z0-9_]*)\s*=\s*["\'][^"\']*["\']',
                r'let\s+([a-zA-Z_][a-zA-Z0-9_]*)\s*=\s*["\'][^"\']*["\']',
                r'const\s+([a-zA-Z_][a-zA-Z0-9_]*)\s*=\s*["\'][^"\']*["\']',
            ]
            
            for pattern in patterns:
                for match in re.finditer(pattern, html, re.IGNORECASE):
                    param_name = match.group(1)
                    if param_name and len(param_name) < 50:
                        if param_name not in params:
                            params[param_name] = ""
            
            url_param_pattern = r'[?&]([a-zA-Z_][a-zA-Z0-9_]*)=([^&"\'>\s]*)'
            for match in re.finditer(url_param_pattern, html):
                param_name = match.group(1)
                param_value = unquote(match.group(2))
                if param_name and len(param_name) < 50:
                    if param_name not in params:
                        params[param_name] = param_value
            
            json_pattern = r'"([a-zA-Z_][a-zA-Z0-9_]*)"\s*:\s*"([^"]*)"'
            for match in re.finditer(json_pattern, html):
                param_name = match.group(1)
                param_value = match.group(2)
                if param_name and len(param_name) < 50 and param_name not in ['url', 'href', 'src', 'type', 'id', 'class']:
                    if param_name not in params:
                        params[param_name] = param_value
        
        except Exception:
            pass
        
        return params
    
    async def _fetch_page(self, url: str) -> Optional[CrawledPage]:
        """获取单个页面"""
        async with self._semaphore:
            try:
                async with self._session.get(
                    url,
                    timeout=aiohttp.ClientTimeout(total=self.timeout),
                    allow_redirects=True
                ) as response:
                    content = await response.text()
                    
                    links = self._extract_links(url, content)
                    forms = self._extract_forms(url, content)
                    url_params = self._extract_parameters_from_url(url)
                    html_params = self._extract_parameters_from_html(content, url)
                    
                    all_params = {**html_params, **url_params}
                    
                    for param_name in all_params.keys():
                        if param_name not in self._all_params:
                            self._all_params[param_name] = set()
                        self._all_params[param_name].add(url)
                    
                    return CrawledPage(
                        url=url,
                        status_code=response.status,
                        content=content,
                        links=links,
                        forms=forms,
                        parameters=all_params
                    )
            except Exception as e:
                return None
    
    async def crawl(self, start_url: str, allowed_domains: List[str] = None, 
                   blocked_domains: List[str] = None) -> List[CrawledPage]:
        """爬取网站"""
        self._visited = set()
        self._results = []
        self._all_params = {}
        self._semaphore = asyncio.Semaphore(self.concurrent)
        
        headers = {
            "User-Agent": self.user_agent,
            "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
            "Accept-Language": "en-US,en;q=0.9",
        }
        
        connector = aiohttp.TCPConnector(ssl=False)
        self._session = aiohttp.ClientSession(headers=headers, connector=connector)
        
        try:
            queue = [(start_url, 0)]
            
            while queue and len(self._results) < self.max_pages:
                url, depth = queue.pop(0)
                
                normalized = self._normalize_url(url)
                if normalized in self._visited:
                    continue
                
                parsed = urlparse(url)
                domain = parsed.netloc
                
                if blocked_domains and domain in blocked_domains:
                    continue
                
                if allowed_domains and domain not in allowed_domains:
                    continue
                
                self._visited.add(normalized)
                
                page = await self._fetch_page(url)
                if page:
                    self._results.append(page)
                    
                    if depth < self.max_depth:
                        for link in page.links:
                            normalized_link = self._normalize_url(link)
                            if normalized_link not in self._visited:
                                link_parsed = urlparse(link)
                                link_domain = link_parsed.netloc
                                
                                if blocked_domains and link_domain in blocked_domains:
                                    continue
                                
                                if allowed_domains and link_domain not in allowed_domains:
                                    continue
                                
                                queue.append((link, depth + 1))
        
        finally:
            await self._session.close()
        
        return self._results
    
    def get_all_urls(self) -> List[str]:
        """获取所有爬取的URL"""
        return [page.url for page in self._results]
    
    def get_all_forms(self) -> List[FormInfo]:
        """获取所有表单"""
        forms = []
        for page in self._results:
            for form in page.forms:
                forms.append(FormInfo(
                    action=form['action'],
                    method=form['method'],
                    inputs=form['inputs']
                ))
        return forms
    
    def get_all_parameters(self) -> Dict[str, List[str]]:
        """获取所有参数及其来源URL"""
        result = {}
        for param_name, urls in self._all_params.items():
            result[param_name] = list(urls)
        return result
    
    def get_unique_parameters(self) -> List[str]:
        """获取所有唯一的参数名"""
        return list(self._all_params.keys())

File: RCE_HawkEye_linux/internal/crawler/crawler.go
package crawler

import (
	"context"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"sync"
	"time"

	"github.com/hbzw/RCE_HawkEye_go/internal/types"
	"github.com/hbzw/RCE_HawkEye_go/internal/utils"
)

type Crawler struct {
	maxDepth     int
	maxPages     int
	timeout      int
	concurrent   int
	userAgent    string
	client       *http.Client
	visited      map[string]bool
	visitedMutex sync.Mutex
	results      []types.CrawledPage
	resultsMutex sync.Mutex
	allowedDomains []string
	blockedDomains []string
	restrictRoot bool
	rootDomain   string
}

type CrawlerOption func(*Crawler)

func WithMaxDepth(depth int) CrawlerOption {
	return func(c *Crawler) {
		c.maxDepth = depth
	}
}

func WithMaxPages(pages int) CrawlerOption {
	return func(c *Crawler) {
		c.maxPages = pages
	}
}

func WithTimeout(timeout int) CrawlerOption {
	return func(c *Crawler) {
		c.timeout = timeout
	}
}

func WithConcurrent(concurrent int) CrawlerOption {
	return func(c *Crawler) {
		c.concurrent = concurrent
	}
}

func WithUserAgent(ua string) CrawlerOption {
	return func(c *Crawler) {
		c.userAgent = ua
	}
}

func WithAllowedDomains(domains []string) CrawlerOption {
	return func(c *Crawler) {
		c.allowedDomains = domains
	}
}

func WithBlockedDomains(domains []string) CrawlerOption {
	return func(c *Crawler) {
		c.blockedDomains = domains
	}
}

func WithRestrictRoot(restrict bool) CrawlerOption {
	return func(c *Crawler) {
		c.restrictRoot = restrict
	}
}

func NewCrawler(opts ...CrawlerOption) *Crawler {
	c := &Crawler{
		maxDepth:     2,
		maxPages:     100,
		timeout:      10,
		concurrent:   5,
		userAgent:    "RCE-HawkEye/1.1.0",
		visited:      make(map[string]bool),
		results:      make([]types.CrawledPage, 0),
		allowedDomains: make([]string, 0),
		blockedDomains: make([]string, 0),
		restrictRoot: true,
	}

	for _, opt := range opts {
		opt(c)
	}

	c.client = &http.Client{
		Timeout: time.Duration(c.timeout) * time.Second,
		CheckRedirect: func(req *http.Request, via []*http.Request) error {
			return http.ErrUseLastResponse
		},
	}

	return c
}

func (c *Crawler) Crawl(ctx context.Context, startURL string) ([]types.CrawledPage, error) {
	normalizedURL := utils.NormalizeTarget(startURL)

	parsed, err := url.Parse(normalizedURL)
	if err != nil {
		return nil, fmt.Errorf("invalid URL: %v", err)
	}

	c.rootDomain = parsed.Hostname()

	if len(c.allowedDomains) == 0 && c.restrictRoot {
		c.allowedDomains = []string{c.rootDomain}
	}

	c.crawlPage(ctx, normalizedURL, 0)

	return c.results, nil
}

func (c *Crawler) crawlPage(ctx context.Context, pageURL string, depth int) {
	c.visitedMutex.Lock()
	if c.visited[pageURL] {
		c.visitedMutex.Unlock()
		return
	}

	if len(c.visited) >= c.maxPages {
		c.visitedMutex.Unlock()
		return
	}

	c.visited[pageURL] = true
	c.visitedMutex.Unlock()

	if depth > c.maxDepth {
		return
	}

	page, links := c.fetchPage(ctx, pageURL)
	if page == nil {
		return
	}

	c.resultsMutex.Lock()
	c.results = append(c.results, *page)
	c.resultsMutex.Unlock()

	for _, link := range links {
		if c.isAllowedDomain(link) {
			go c.crawlPage(ctx, link, depth+1)
		}
	}
}

func (c *Crawler) fetchPage(ctx context.Context, pageURL string) (*types.CrawledPage, []string) {
	req, err := http.NewRequestWithContext(ctx, "GET", pageURL, nil)
	if err != nil {
		return nil, nil
	}

	req.Header.Set("User-Agent", c.userAgent)
	req.Header.Set("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8")

	resp, err := c.client.Do(req)
	if err != nil {
		return nil, nil
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		return nil, nil
	}

	contentType := resp.Header.Get("Content-Type")
	if !strings.Contains(contentType, "text/html") && !strings.Contains(contentType, "application/xhtml+xml") {
		return nil, nil
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, nil
	}

	content := string(body)
	links := c.extractLinks(pageURL, content)
	forms := c.extractForms(content)
	params := c.extractParameters(pageURL, content)

	page := &types.CrawledPage{
		URL:        pageURL,
		StatusCode: resp.StatusCode,
		Content:    content,
		Links:      links,
		Forms:      forms,
		Parameters: params,
	}

	return page, links
}

func (c *Crawler) extractLinks(baseURL, content string) []string {
	var links []string
	seen := make(map[string]bool)

	linkPatterns := []string{
		`href=["']([^"']+)["']`,
		`src=["']([^"']+)["']`,
		`action=["']([^"']+)["']`,
	}

	for _, pattern := range linkPatterns {
		re := regexp.MustCompile(pattern)
		matches := re.FindAllStringSubmatch(content, -1)

		for _, match := range matches {
			if len(match) > 1 {
				link := match[1]
				absoluteURL := c.resolveURL(baseURL, link)

				if absoluteURL != "" && !seen[absoluteURL] && c.isValidLink(absoluteURL) {
					seen[absoluteURL] = true
					links = append(links, absoluteURL)
				}
			}
		}
	}

	return links
}

func (c *Crawler) extractForms(content string) []map[string]interface{} {
	var forms []map[string]interface{}

	formPattern := regexp.MustCompile(`(?i)<form[^>]*>(.*?)</form>`)
	inputPattern := regexp.MustCompile(`(?i)<input[^>]*name=["']([^"']+)["'][^>]*(?:value=["']([^"']*)["'])?`)

	formMatches := formPattern.FindAllStringSubmatch(content, -1)
	for _, formMatch := range formMatches {
		formHTML := formMatch[0]

		form := map[string]interface{}{
			"action": "",
			"method": "GET",
			"inputs": make(map[string]string),
		}

		actionMatch := regexp.MustCompile(`(?i)action=["']([^"']*)["']`).FindStringSubmatch(formHTML)
		if len(actionMatch) > 1 {
			form["action"] = actionMatch[1]
		}

		methodMatch := regexp.MustCompile(`(?i)method=["']([^"']*)["']`).FindStringSubmatch(formHTML)
		if len(methodMatch) > 1 {
			form["method"] = strings.ToUpper(methodMatch[1])
		}

		inputMatches := inputPattern.FindAllStringSubmatch(formHTML, -1)
		inputs := make(map[string]string)
		for _, inputMatch := range inputMatches {
			if len(inputMatch) > 1 {
				name := inputMatch[1]
				value := ""
				if len(inputMatch) > 2 {
					value = inputMatch[2]
				}
				inputs[name] = value
			}
		}
		form["inputs"] = inputs

		forms = append(forms, form)
	}

	return forms
}

func (c *Crawler) extractParameters(pageURL, content string) map[string]string {
	params := make(map[string]string)

	parsed, err := url.Parse(pageURL)
	if err == nil {
		for key, values := range parsed.Query() {
			if len(values) > 0 {
				params[key] = values[0]
			}
		}
	}

	paramPatterns := []string{
		`var\s+(\w+)\s*=\s*["']([^"']+)["']`,
		`const\s+(\w+)\s*=\s*["']([^"']+)["']`,
		`let\s+(\w+)\s*=\s*["']([^"']+)["']`,
	}

	for _, pattern := range paramPatterns {
		re := regexp.MustCompile(pattern)
		matches := re.FindAllStringSubmatch(content, -1)
		for _, match := range matches {
			if len(match) > 2 {
				params[match[1]] = match[2]
			}
		}
	}

	apiPatterns := []string{
		`url\s*:\s*["']([^"']+)["']`,
		`endpoint\s*:\s*["']([^"']+)["']`,
		`api\s*:\s*["']([^"']+)["']`,
		`fetch\s*\(\s*["']([^"']+)["']`,
		`axios\s*\.\s*\w+\s*\(\s*["']([^"']+)["']`,
		`\$\.(?:get|post|ajax)\s*\(\s*["']([^"']+)["']`,
		`XMLHttpRequest.*open\s*\(\s*["'][^"']*["']\s*,\s*["']([^"']+)["']`,
	}
	
	for _, pattern := range apiPatterns {
		re := regexp.MustCompile(pattern)
		matches := re.FindAllStringSubmatch(content, -1)
		for _, match := range matches {
			if len(match) > 1 && strings.Contains(match[1], "?") {
				apiURL := match[1]
				if parsed, err := url.Parse(apiURL); err == nil {
					for key, values := range parsed.Query() {
						if len(values) > 0 {
							params[key] = values[0]
						}
					}
				}
			}
		}
	}

	dataAttrPattern := `data-[\w-]+\s*=\s*["']([^"']+)["']`
	re := regexp.MustCompile(dataAttrPattern)
	matches := re.FindAllStringSubmatch(content, -1)
	for _, match := range matches {
		if len(match) > 1 && len(match[1]) > 0 && len(match[1]) < 100 {
			params["data_attr"] = match[1]
		}
	}

	ajaxPatterns := []string{
		`type\s*:\s*["'](POST|GET|PUT|DELETE)["']`,
		`method\s*:\s*["'](POST|GET|PUT|DELETE)["']`,
	}
	
	for _, pattern := range ajaxPatterns {
		re := regexp.MustCompile(pattern)
		if re.MatchString(content) {
			params["_ajax_method"] = "detected"
			break
		}
	}

	jsonPattern := `["'](\w+)["']\s*:\s*["']([^"']+)["']`
	re = regexp.MustCompile(jsonPattern)
	matches = re.FindAllStringSubmatch(content, -1)
	for _, match := range matches {
		if len(match) > 2 && len(match[2]) > 0 && len(match[2]) < 100 {
			key := match[1]
			if !strings.HasPrefix(key, "_") && !strings.Contains(key, "token") {
				params[key] = match[2]
			}
		}
	}

	hiddenInputPattern := `<input[^>]*type=["']hidden["'][^>]*name=["'](\w+)["'][^>]*value=["']([^"']*)["']`
	re = regexp.MustCompile(hiddenInputPattern)
	matches = re.FindAllStringSubmatch(content, -1)
	for _, match := range matches {
		if len(match) > 2 {
			params[match[1]] = match[2]
		}
	}

	textareaPattern := `<textarea[^>]*name=["'](\w+)["'][^>]*>([^<]*)</textarea>`
	re = regexp.MustCompile(textareaPattern)
	matches = re.FindAllStringSubmatch(content, -1)
	for _, match := range matches {
		if len(match) > 2 {
			params[match[1]] = match[2]
		}
	}

	selectPattern := `<select[^>]*name=["'](\w+)["']`
	re = regexp.MustCompile(selectPattern)
	matches = re.FindAllStringSubmatch(content, -1)
	for _, match := range matches {
		if len(match) > 1 {
			params[match[1]] = ""
		}
	}

	return params
}

func (c *Crawler) resolveURL(baseURL, link string) string {
	if link == "" {
		return ""
	}

	link = strings.TrimSpace(link)

	if strings.HasPrefix(link, "#") ||
		strings.HasPrefix(link, "javascript:") ||
		strings.HasPrefix(link, "mailto:") ||
		strings.HasPrefix(link, "tel:") ||
		strings.HasPrefix(link, "data:") {
		return ""
	}

	base, err := url.Parse(baseURL)
	if err != nil {
		return ""
	}

	ref, err := url.Parse(link)
	if err != nil {
		return ""
	}

	resolved := base.ResolveReference(ref)

	if resolved.Scheme != "http" && resolved.Scheme != "https" {
		return ""
	}

	return resolved.String()
}

func (c *Crawler) isValidLink(link string) bool {
	excludedExtensions := []string{
		".jpg", ".jpeg", ".png", ".gif", ".bmp", ".svg", ".ico",
		".css", ".woff", ".woff2", ".ttf", ".eot",
		".pdf", ".doc", ".docx", ".xls", ".xlsx", ".ppt", ".pptx",
		".zip", ".rar", ".tar", ".gz", ".7z",
		".mp3", ".mp4", ".avi", ".mov", ".wmv", ".flv",
		".xml", ".json", ".rss", ".atom",
	}

	linkLower := strings.ToLower(link)
	for _, ext := range excludedExtensions {
		if strings.HasSuffix(linkLower, ext) {
			return false
		}
	}

	return true
}

func (c *Crawler) isAllowedDomain(link string) bool {
	parsed, err := url.Parse(link)
	if err != nil {
		return false
	}

	domain := parsed.Hostname()

	for _, blocked := range c.blockedDomains {
		if domain == blocked || strings.HasSuffix(domain, "."+blocked) {
			return false
		}
	}

	if len(c.allowedDomains) == 0 {
		return true
	}

	for _, allowed := range c.allowedDomains {
		if domain == allowed || strings.HasSuffix(domain, "."+allowed) {
			return true
		}
	}

	return false
}

func (c *Crawler) GetResults() []types.CrawledPage {
	return c.results
}

func (c *Crawler) GetAllURLs() []string {
	var urls []string
	for _, page := range c.results {
		urls = append(urls, page.URL)
	}
	return urls
}

func (c *Crawler) GetAllForms() []map[string]interface{} {
	var allForms []map[string]interface{}
	for _, page := range c.results {
		allForms = append(allForms, page.Forms...)
	}
	return allForms
}

func (c *Crawler) GetAllParameters() map[string]string {
	allParams := make(map[string]string)
	for _, page := range c.results {
		for k, v := range page.Parameters {
			allParams[k] = v
		}
	}
	return allParams
}

func (c *Crawler) GetStatistics() map[string]int {
	return map[string]int{
		"total_pages":   len(c.results),
		"total_links":   len(c.GetAllURLs()),
		"total_forms":   len(c.GetAllForms()),
		"total_params":  len(c.GetAllParameters()),
	}
}
